File: apps/web/src/lib/api/reactions.ts
/**
 * Reactions API client (Wave 8 Stream C-2).
 *
 * Wraps `POST /api/v1/reactions` and `GET /api/v1/reactions/summary`.
 * The wire shape is kept aligned with the OpenAPI-generated
 * `ReactionSummaryResponse` / `SetReactionResponse` / `SummaryBatchResponse`
 * components. Article and source detail responses already embed the same
 * `reaction_summary` field, so consumers can read it directly off the entity.
 */

import { apiClient } from "@/lib/api/client";

// ----------------------------------------------------------------------------
// Wire types
// ----------------------------------------------------------------------------

export type ReactionTargetType = "article" | "source";
export type ReactionKind = "like" | "dislike";

export interface ReactionSummary {
	likes: number;
	dislikes: number;
	score: number;
	/**
	 * `"like"` / `"dislike"` / undefined. The backend omits this field for
	 * anonymous callers. Treat undefined as "no reaction yet".
	 */
	my_kind?: ReactionKind | null;
}

export interface SetReactionInput {
	targetType: ReactionTargetType;
	targetId: string;
	/** `null` clears any existing reaction. */
	kind: ReactionKind | null;
}

export interface SetReactionResponse {
	summary: ReactionSummary;
}

export interface SummaryBatchInput {
	targetType: ReactionTargetType;
	/** Up to 100 target ids per call. The caller is responsible for chunking. */
	targetIds: string[];
}

export interface SummaryBatchResponse {
	summaries: Record<string, ReactionSummary>;
}

export const REACTION_BATCH_LIMIT = 100;

// ----------------------------------------------------------------------------
// Runtime validators
// ----------------------------------------------------------------------------

const TARGET_TYPES: readonly ReactionTargetType[] = ["article", "source"];
const KINDS: readonly ReactionKind[] = ["like", "dislike"];

function isReactionKind(value: unknown): value is ReactionKind {
	return (
		typeof value === "string" && (KINDS as readonly string[]).includes(value)
	);
}

function asRecord(value: unknown, path: string): Record<string, unknown> {
	if (typeof value !== "object" || value === null || Array.isArray(value)) {
		throw new Error(`${path}: expected object, got ${typeof value}`);
	}
	return value as Record<string, unknown>;
}

function asNumber(value: unknown, path: string): number {
	if (typeof value !== "number" || Number.isNaN(value)) {
		throw new Error(`${path}: expected number, got ${typeof value}`);
	}
	return value;
}

export function assertReactionSummary(
	value: unknown,
	path = "reactionSummary",
): asserts value is ReactionSummary {
	const record = asRecord(value, path);
	asNumber(record.likes, `${path}.likes`);
	asNumber(record.dislikes, `${path}.dislikes`);
	asNumber(record.score, `${path}.score`);

	if (record.my_kind !== undefined && record.my_kind !== null) {
		if (!isReactionKind(record.my_kind)) {
			throw new Error(
				`${path}.my_kind: expected "like" | "dislike" | null | undefined, got ${String(record.my_kind)}`,
			);
		}
	}
}

export function assertSetReactionResponse(
	value: unknown,
): asserts value is SetReactionResponse {
	const record = asRecord(value, "setReactionResponse");
	assertReactionSummary(record.summary, "setReactionResponse.summary");
}

export function assertSummaryBatchResponse(
	value: unknown,
): asserts value is SummaryBatchResponse {
	const record = asRecord(value, "summaryBatchResponse");
	const summaries = asRecord(
		record.summaries,
		"summaryBatchResponse.summaries",
	);
	for (const [id, summary] of Object.entries(summaries)) {
		assertReactionSummary(summary, `summaryBatchResponse.summaries["${id}"]`);
	}
}

// ----------------------------------------------------------------------------
// Helpers
// ----------------------------------------------------------------------------

/**
 * The empty/zero summary. Use this when the backend has no record for a target
 * yet (e.g. a freshly created article that no one has reacted to).
 */
export const EMPTY_REACTION_SUMMARY: ReactionSummary = {
	likes: 0,
	dislikes: 0,
	score: 0,
};

export function isReactionTargetType(
	value: unknown,
): value is ReactionTargetType {
	return (
		typeof value === "string" &&
		(TARGET_TYPES as readonly string[]).includes(value)
	);
}

// ----------------------------------------------------------------------------
// Endpoints
// ----------------------------------------------------------------------------

export async function setReaction(
	input: SetReactionInput,
): Promise<SetReactionResponse> {
	const body = {
		target_type: input.targetType,
		target_id: input.targetId,
		kind: input.kind,
	};
	return apiClient.post<SetReactionResponse>(
		"/api/v1/reactions",
		body,
		assertSetReactionResponse,
	);
}

export async function getReactionSummariesBatch(
	input: SummaryBatchInput,
): Promise<SummaryBatchResponse> {
	if (input.targetIds.length === 0) {
		return { summaries: {} };
	}
	if (input.targetIds.length > REACTION_BATCH_LIMIT) {
		throw new Error(
			`getReactionSummariesBatch: targetIds capped at ${REACTION_BATCH_LIMIT}, got ${input.targetIds.length}`,
		);
	}

	const params = new URLSearchParams();
	params.set("target_type", input.targetType);
	for (const id of input.targetIds) {
		params.append("target_ids[]", id);
	}

	return apiClient.get<SummaryBatchResponse>(
		`/api/v1/reactions/summary?${params.toString()}`,
		assertSummaryBatchResponse,
	);
}

File: crates/law-eye-api/src/routes/reactions.rs
//! Wave-8 Stream C-1: public reactions endpoints.
//!
//! `POST   /api/v1/reactions`               — set/clear the calling user's reaction
//! `GET    /api/v1/reactions/summary`       — batch summary lookup for N target ids
//!
//! Login required (the protected_api router enforces RequireAuth one level up).
//! Both endpoints serialise the canonical `ReactionSummaryResponse` envelope so
//! the frontend hook can share a single shape between detail and batch flows.

use std::collections::HashMap;
use std::str::FromStr;

use axum::{
    extract::{Query, State},
    routing::{get, post},
    Json, Router,
};
use serde::{Deserialize, Serialize};
use utoipa::{IntoParams, ToSchema};
use uuid::Uuid;

use law_eye_core::{
    ReactionKind, ReactionService as _, ReactionSummary, ReactionTarget,
};

use crate::auth::AuthSession;
use crate::state::AppState;
use crate::{ApiJson, ApiResult, AppError};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", post(set_reaction))
        .route("/summary", get(get_summary_batch))
}

#[derive(Debug, Clone, Serialize, ToSchema, PartialEq, Eq)]
pub struct ReactionSummaryResponse {
    pub likes: i64,
    pub dislikes: i64,
    pub score: i64,
    /// `"like"` / `"dislike"` / null. Present only when the request has an
    /// authenticated viewer; the field is omitted otherwise so anonymous
    /// callers never accidentally see another user's state.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub my_kind: Option<String>,
}

impl From<ReactionSummary> for ReactionSummaryResponse {
    fn from(summary: ReactionSummary) -> Self {
        Self {
            likes: summary.likes,
            dislikes: summary.dislikes,
            score: summary.score,
            my_kind: summary.my_kind.map(|k| k.as_str().to_string()),
        }
    }
}

#[derive(Debug, Clone, Deserialize, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct SetReactionRequest {
    /// `"article"` or `"source"`.
    pub target_type: String,
    pub target_id: Uuid,
    /// `"like"` / `"dislike"` / null. `null` clears any existing reaction.
    pub kind: Option<String>,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct SetReactionResponse {
    pub summary: ReactionSummaryResponse,
}

#[derive(Debug, Clone, Deserialize, IntoParams)]
pub struct SummaryQuery {
    /// `"article"` or `"source"`.
    pub target_type: String,
    /// Repeated `target_ids[]=<uuid>` query parameter; up to 100 ids per call.
    #[serde(default, rename = "target_ids[]")]
    pub target_ids: Vec<Uuid>,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct SummaryBatchResponse {
    pub summaries: HashMap<String, ReactionSummaryResponse>,
}

const MAX_BATCH_TARGET_IDS: usize = 100;

fn parse_target_type(raw: &str) -> Result<ReactionTarget, AppError> {
    ReactionTarget::from_str(raw)
        .map_err(|e| AppError::validation_with_code("INVALID_TARGET_TYPE", e))
}

fn parse_kind(raw: Option<&str>) -> Result<Option<ReactionKind>, AppError> {
    match raw {
        None => Ok(None),
        Some(value) => ReactionKind::from_str(value)
            .map(Some)
            .map_err(|e| AppError::validation_with_code("INVALID_REACTION_KIND", e)),
    }
}

#[utoipa::path(
    post,
    path = "/api/v1/reactions",
    request_body = SetReactionRequest,
    security(("session" = [])),
    responses(
        (status = 200, description = "Updated reaction summary", body = SetReactionResponse),
        (status = 400, description = "Invalid payload"),
        (status = 401, description = "Authentication required"),
        (status = 404, description = "Target article or source not found")
    ),
    tag = "reactions"
)]
pub async fn set_reaction(
    State(state): State<AppState>,
    auth_session: AuthSession,
    ApiJson(req): ApiJson<SetReactionRequest>,
) -> ApiResult<Json<SetReactionResponse>> {
    let user = auth_session
        .user
        .ok_or_else(|| AppError::unauthorized("Authentication required"))?;

    let target_type = parse_target_type(&req.target_type)?;
    let kind = parse_kind(req.kind.as_deref())?;

    let summary = state
        .reaction_service
        .set_reaction(user.tenant_id, user.id, target_type, req.target_id, kind)
        .await
        .map_err(AppError::from)?;

    Ok(Json(SetReactionResponse {
        summary: summary.into(),
    }))
}

#[utoipa::path(
    get,
    path = "/api/v1/reactions/summary",
    params(SummaryQuery),
    security(("session" = [])),
    responses(
        (status = 200, description = "Reaction summaries keyed by target id", body = SummaryBatchResponse),
        (status = 400, description = "Invalid params or too many target ids"),
        (status = 401, description = "Authentication required")
    ),
    tag = "reactions"
)]
pub async fn get_summary_batch(
    State(state): State<AppState>,
    auth_session: AuthSession,
    Query(params): Query<SummaryQuery>,
) -> ApiResult<Json<SummaryBatchResponse>> {
    let user = auth_session
        .user
        .ok_or_else(|| AppError::unauthorized("Authentication required"))?;

    let target_type = parse_target_type(&params.target_type)?;
    if params.target_ids.len() > MAX_BATCH_TARGET_IDS {
        return Err(AppError::validation_with_code(
            "TOO_MANY_TARGETS",
            format!(
                "target_ids[] is capped at {} entries",
                MAX_BATCH_TARGET_IDS
            ),
        ));
    }

    let map = state
        .reaction_service
        .get_summaries_batch(
            user.tenant_id,
            target_type,
            &params.target_ids,
            Some(user.id),
        )
        .await
        .map_err(AppError::from)?;

    let summaries: HashMap<String, ReactionSummaryResponse> = map
        .into_iter()
        .map(|(id, summary)| (id.to_string(), summary.into()))
        .collect();

    Ok(Json(SummaryBatchResponse { summaries }))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_target_type_accepts_known_values() {
        assert_eq!(parse_target_type("article").unwrap(), ReactionTarget::Article);
        assert_eq!(parse_target_type("SOURCE").unwrap(), ReactionTarget::Source);
        let err = parse_target_type("comment").unwrap_err();
        assert_eq!(err.body.code.as_deref(), Some("INVALID_TARGET_TYPE"));
    }

    #[test]
    fn parse_kind_handles_optional_payload() {
        assert!(parse_kind(None).unwrap().is_none());
        assert_eq!(parse_kind(Some("like")).unwrap(), Some(ReactionKind::Like));
        let err = parse_kind(Some("meh")).unwrap_err();
        assert_eq!(err.body.code.as_deref(), Some("INVALID_REACTION_KIND"));
    }

    #[test]
    fn summary_response_omits_my_kind_when_absent() {
        let response = ReactionSummaryResponse::from(ReactionSummary::new(2, 1, None));
        let json = serde_json::to_value(&response).unwrap();
        assert_eq!(json.get("likes").and_then(|v| v.as_i64()), Some(2));
        assert_eq!(json.get("score").and_then(|v| v.as_i64()), Some(1));
        assert!(json.get("my_kind").is_none());
    }
}
